pkg/util/validation/limits.go

```diff
@@ -114,184 +111,153 @@ func (l *Limits) UnmarshalYAML(unmarshal func(interface{}) error) error {
 // find a nicer way I'm afraid.
 var defaultLimits *Limits
 
+// TenantLimits is a function that returns limits for given tenant, or
+// nil, if there are no tenant-specific limits.
+type TenantLimits func(userID string) *Limits
+
 // Overrides periodically fetch a set of per-user overrides, and provides convenience
 // functions for fetching the correct value.
 type Overrides struct {
-	overridesManager *OverridesManager
+	defaultLimits *Limits
+	tenantLimits  TenantLimits
 }
 
 // NewOverrides makes a new Overrides.
 // We store the supplied limits in a global variable to ensure per-tenant limits
 // are defaulted to those values.  As such, the last call to NewOverrides will
 // become the new global defaults.
-func NewOverrides(defaults Limits) (*Overrides, error) {
-	defaultLimits = &defaults
```

Comment:
This needs to be retained since it sets the default values to the global variable `defaultLimits` without which the limits which are not set in overrides would be set to zero values. Try removing line https://github.com/cortexproject/cortex/blob/d6edf4b2cacf8a19e026b8d7e86144491143334a/pkg/util/override_test.go#L101 if you want to see the impact.

Maybe to avoid this from happening again, do you want to add an end to end test somewhere which uses the same global variable from this package?

Reply:
Oh, you're correct! I broke it :-( Fixing now. I'll add a test for this, to make sure it doesn't get broken again.

Reply:
Thanks for catching this!

Reply:
Done.

Reply:
Thinking about it... perhaps we should just have SetDefaultLimits method in that package instead?

Reply:
I've moved it to separate function now, called `SetDefaultLimitsForYAMLUnmarshalling`.

Reply:
I think it would be better calling `SetDefaultLimitsForYAMLUnmarshalling` from `limits.NewOverrides` itself. It would be easy to miss calling it.

Reply:
I prefer code to be explicit, rather than relying on hidden side-effects like that. It is only used in single place in Cortex (not counting tests, which don't rely on it) and I've added call to that place. Of course, we need to add that to Loki, when it is updated to use this code, but there are more changes related to this PR that will be required in Loki.

Reply:
My perspective:
- Looking at the func name `NewOverrides` I would expect it just create new overrides, without touching the global variable storing default limits. For this reason, I would suggest to be explicit and call the func to set default limits outside `NewOverrides`
- The name `SetDefaultLimitsForYAMLUnmarshalling()` looks a bit weird to me, because it doesn't have logic which is specific of the fact that the default limits come from YAML unmarshalling. I would simply call it `SetDefaultLimits()` and then explain - with a code comment in `modules.go` - that we're calling it to set default limits from value read from YAML

Reply:
> * The name `SetDefaultLimitsForYAMLUnmarshalling()` looks a bit weird to me

That was my goal... to make it look weird to make sure people take a deeper look before using/changing it. It's not setting default limits (which `SetDefaultLimits` would suggest) -- default limits are passed to the `NewOverrides` as a first argument. This function is setting default limits that are ONLY USED during YAML unmarshal, nowhere else. I think calling it just `SetDefaultLimits` would be misleading.

Reply:
Ok, thanks for the explanation. I'm good with the current naming and design then.